staging/src/k8s.io/apiserver/pkg/server/filters/compression.go

```diff
@@ -0,0 +1,147 @@
+package filters
+
+import (
+	"compress/gzip"
+	"compress/zlib"
+	"errors"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/emicklei/go-restful"
+)
+
+type Compressor interface {
+	io.WriteCloser
+	Flush() error
+}
+
+const (
+	header_AcceptEncoding  = "Accept-Encoding"
+	header_ContentEncoding = "Content-Encoding"
+
+	encoding_gzip    = "gzip"
+	encoding_deflate = "deflate"
+)
+
+func WithCompression(handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		//don't compress watches
+		if req.URL.Query().Get("watch") == "true" {
+			handler.ServeHTTP(w, req)
+			return
+		}
+		wantsCompression, encoding := wantsCompressedResponse(req)
+		if wantsCompression {
+			compressionWriter := NewCompressionResponseWriter(w, encoding)
+			compressionWriter.Header().Set("Content-Encoding", encoding)
+			handler.ServeHTTP(compressionWriter, req)
+		} else {
+			handler.ServeHTTP(w, req)
+		}
+	})
+
+}
+
+// WantsCompressedResponse reads the Accept-Encoding header to see if and which encoding is requested.
+func wantsCompressedResponse(httpRequest *http.Request) (bool, string) {
+	if watch := httpRequest.URL.Query().Get("watch"); watch == "true" || watch == "1" {
+		return false, ""
+	}
+	header := httpRequest.Header.Get(header_AcceptEncoding)
+	gi := strings.Index(header, encoding_gzip)
+	zi := strings.Index(header, encoding_deflate)
+	// use in order of appearance
+	if gi == -1 {
+		return zi != -1, encoding_deflate
+	} else if zi == -1 {
+		return gi != -1, encoding_gzip
+	} else {
+		if gi < zi {
+			return true, encoding_gzip
+		}
+		return true, encoding_deflate
+	}
+}
+
+type compressionResponseWriter struct {
+	writer     http.ResponseWriter
+	compressor Compressor
+	encoding   string
+}
+
+func NewCompressionResponseWriter(w http.ResponseWriter, encoding string) *compressionResponseWriter {
+	var compressor Compressor
+	switch encoding {
+	case encoding_gzip:
+		compressor = gzip.NewWriter(w)
+	case encoding_deflate:
+		compressor = zlib.NewWriter(w)
+	default:
+		panic(encoding + " not a valid compression type")
```

Comment:
This shouldn't be here at all; `wantsCompression` simply returns false if the user specifies a compression type that doesn't exactly match `gzip` or `deflate`. I'll fix this in my next commit